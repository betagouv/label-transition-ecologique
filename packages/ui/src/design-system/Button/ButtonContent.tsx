import {Icon, IconSize} from '@design-system/Icon';
import {buttonThemeClassnames} from './theme';
import {ButtonContentProps, ButtonSize, ButtonVariant} from './types';
import classNames from 'classnames';

const getIconSize = (
  size: ButtonSize,
  variant: ButtonVariant
): IconSize | undefined => {
  // Les sizes des icÃ´nes ne matchent pas celles des boutons
  switch (size) {
    case 'xs':
      return variant === 'underlined' ? 'xs' : 'sm';
    case 'sm':
      return variant === 'underlined' ? 'sm' : 'md';
    case 'md':
      return 'md';
    case 'xl':
      return 'lg';
  }
};

/** Affiche le contenu d'un bouton */
const ButtonContent = ({
  variant = 'primary',
  size = 'md',
  icon,
  children,
  disabled,
}: ButtonContentProps) => {
  const buttonState = disabled ? 'disabled' : 'default';

  return (
    <>
      {!!icon && (
        <Icon
          icon={icon}
          size={getIconSize(size, variant)}
          className={classNames(
            buttonThemeClassnames[variant][buttonState].icon,
            {
              'mb-1': variant === 'underlined' && size === 'sm',
              'mb-0.5':
                variant === 'underlined' && (size === 'xs' || size === 'md'),
            }
          )}
        />
      )}

      {!!children && children}
    </>
  );
};

export default ButtonContent;
