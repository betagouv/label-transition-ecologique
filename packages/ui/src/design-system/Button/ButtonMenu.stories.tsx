import {Meta, StoryObj} from '@storybook/react';

import {ButtonMenu} from './ButtonMenu';

const meta: Meta<typeof ButtonMenu> = {
  component: ButtonMenu,
  argTypes: {},
  args: {
    icon: 'equalizer-fill',
    notification: {
      number: 2,
    },
    children: (
      <div className="!w-44 text-sm">
        Lorem ipsum dolor sit amet consectetur adipisicing elit. Laudantium,
        dicta natus, ea voluptates enim nobis aspernatur vero debitis
        repellendus cum temporibus, praesentium neque dolore.
      </div>
    ),
  },
};

export default meta;

type Story = StoryObj<typeof ButtonMenu>;

export const Default: Story = {
  render: args => {
    return (
      <div className="mb-52">
        <ButtonMenu {...args} />
      </div>
    );
  },
};
