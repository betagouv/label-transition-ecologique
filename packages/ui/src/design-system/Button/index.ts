export * from './Button';
export * from './NotificationButton';
