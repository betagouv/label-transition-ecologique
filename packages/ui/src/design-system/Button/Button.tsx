import {Ref, forwardRef} from 'react';
import classNames from 'classnames';

import {ButtonHTMLProps, isAnchor} from 'utils/types';
import {buttonSizeClassnames, buttonThemeClassnames} from './theme';
import ButtonContent from './ButtonContent';
import {ButtonContentProps, ButtonProps} from './types';

/**
 * Composant bouton par défaut, ayant pour props toutes les props habituelles d'un button tag.
 * */
// On déstructure toutes les props rajoutées qui ne sont pas des props des tags HTML <button> ou <a>
// Ce qui nous permet de ne donner que les props restantes natives au tag HTML
export const Button = forwardRef(
  (
    {
      children,
      disabled = false,
      variant = 'primary',
      size = 'md',
      className,
      icon,
      iconPosition,
      external,
      ...props
    }: ButtonProps,
    ref?: Ref<HTMLButtonElement | HTMLAnchorElement>
  ) => {
    const isIconButton = !children;

    const buttonState = disabled ? 'disabled' : 'default';

    const {text, background, border} =
      buttonThemeClassnames[variant][buttonState];

    const sizeClassName =
      buttonSizeClassnames[size][isIconButton ? 'iconButton' : 'textButton'];

    const buttonClassname = classNames(
      'w-fit flex items-center border-solid group',
      {
        // Layout du bouton
        'gap-1': size === 'xs' || (size === 'sm' && variant === 'underlined'),
        'gap-2':
          (size === 'sm' && variant !== 'underlined') ||
          size === 'md' ||
          size === 'xl',
        // Styles du curseur
        'cursor-pointer': !disabled,
        'cursor-not-allowed': disabled,
        // Bordures et polices
        'rounded-lg border font-bold': variant !== 'underlined',
        'border-b border-t-0 border-x-0 !p-px font-medium':
          variant === 'underlined',
        'hover:border-b-2 hover:!pb-0': variant === 'underlined' && !disabled,
      },
      text,
      background,
      border,
      sizeClassName
    );

    /** Reconstitution des props données au contenu du bouton */
    const buttonContentProps: ButtonContentProps = {
      children,
      variant,
      size,
      icon,
      disabled,
    };

    const isButton = !isAnchor(props);

    /** On affiche un bouton par défaut */
    if (isButton) {
      // On réintegre la prop disabled qui a été déstructurée plus haut
      const buttonProps = {...props, disabled} as ButtonHTMLProps;
      return (
        <button
          ref={ref as Ref<HTMLButtonElement>}
          {...buttonProps}
          className={classNames(
            buttonClassname,
            {'flex-row-reverse': iconPosition === 'right'},
            className
          )}
          onClick={evt => {
            evt.stopPropagation();
            buttonProps.onClick?.(evt);
          }}
        >
          <ButtonContent {...buttonContentProps} />
        </button>
      );

      /** Ou bien une ancre si un lien href est donné (cf ./types.ts) */
    } else {
      const anchorProps = props;
      const openInNewTab = external || props.target === '_blank';
      return (
        <a
          {...anchorProps}
          ref={ref as Ref<HTMLAnchorElement>}
          // bg-none permet d'effacer un style dsfr appliqué à la balise <a/>
          // after:hidden supprime l'icône external par défaut du dsfr
          className={classNames(
            'bg-none after:hidden',
            buttonClassname,
            {
              'flex-row-reverse':
                iconPosition === 'right' ||
                (openInNewTab && iconPosition !== 'left'),
            },
            className
          )}
          target={openInNewTab ? '_blank' : anchorProps.target}
          rel={openInNewTab ? 'noreferrer noopener' : anchorProps.rel}
          onClick={evt => {
            evt.stopPropagation();
            if (disabled) evt.preventDefault();
            else anchorProps.onClick?.(evt);
          }}
        >
          <ButtonContent
            {...buttonContentProps}
            icon={openInNewTab ? 'external-link-line' : icon}
          />
        </a>
      );
    }
  }
);
