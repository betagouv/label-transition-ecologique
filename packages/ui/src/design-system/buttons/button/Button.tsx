import classNames from 'classnames';
import {buttonThemeClassnames} from '../utils';
import ButtonContent from './ButtonContent';
import {ButtonProps, isAnchorButton, isDefaultButton} from '../types';

/**
 * Composant bouton par défaut, ayant pour props toutes les props habituelles d'un button tag.
 * */
export const Button = (props: ButtonProps) => {
  const {
    children,
    disabled = false,
    variant = 'primary',
    size = 'md',
    className,
  } = props;

  const isIconButton = !children;
  const {text, background, border} = buttonThemeClassnames[variant];

  const buttonClassname = classNames(
    'rounded-lg w-fit border border-solid font-bold group',
    {
      // Format du bouton en fonction de la taille choisie
      'text-xs py-2': size === 'xs',
      'text-sm py-3': size === 'sm',
      'text-base py-3.5': size === 'md',
      'text-2xl py-5': size === 'xl',

      // Padding en fonction de la taille choisie
      'px-5':
        (size === 'xs' && !isIconButton) || (size === 'xl' && isIconButton),
      'px-6': size !== 'xs' && !isIconButton,

      // Padding sur les boutons ne contenant qu'une icône
      'px-2': size === 'xs' && isIconButton,
      'px-3': size === 'sm' && isIconButton,
      'px-3.5': size === 'md' && isIconButton,

      // Styles du curseur
      'cursor-pointer': !disabled,
      'cursor-not-allowed': disabled,
    },
    text,
    background,
    border
  );

  if (isDefaultButton(props)) {
    return (
      <button {...props} className={classNames(buttonClassname, className)}>
        <ButtonContent {...{...props, children, disabled, variant, size}} />
      </button>
    );
  } else if (isAnchorButton(props)) {
    const {external = false} = props;

    return (
      <a
        {...props}
        // bg-none permet d'effacer un style dsfr appliqué à la balise <a/>
        // after:hidden supprime l'icône external par défaut du dsfr
        className={classNames(
          'inline-block w-fit bg-none after:hidden',
          buttonClassname,
          className
        )}
        target={external ? '_blank' : props.target}
        rel={external ? 'noreferrer noopener' : props.rel}
        onClick={evt => {
          if (disabled) evt.preventDefault();
        }}
      >
        <ButtonContent
          {...{...props, children, disabled, variant, size}}
          external={external || props.target === '_blank'}
        />
      </a>
    );
  }
};
