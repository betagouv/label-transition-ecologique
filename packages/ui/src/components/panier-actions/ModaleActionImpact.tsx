import {Badge} from '@design-system/Badge';
import {Button} from '@design-system/Button';
import {Modal, ModalFooter, ModalFooterSection} from '@design-system/Modal';
// import {Icon} from '@design-system/Icon';
import {NiveauBudget} from './NiveauBudget';
import {valeurToBadge} from './utils';
import {ModaleActionImpactProps} from './types';
import classNames from 'classnames';

/**
 * Modale action à impact du panier d'actions
 */

export const ModaleActionImpact = ({
  children,
  titre,
  thematiques,
  complexite,
  budget,
  description,
  nbCollectivitesEnCours,
  nbCollectivitesRealise,
  statut,
  panier,
  onToggleSelected,
  onUpdateStatus,
}: ModaleActionImpactProps) => {
  return (
    <Modal
      size="lg"
      render={() => (
        <div>
          <div className="flex justify-between mb-6 mt-4">
            {/* Catégorie */}
            <div className="text-primary-8 text-lg font-bold flex divide-x-2 divide-x-primary-3 gap-x-3">
              {thematiques.map(them => (
                <div key={them.id} className="pl-3 first-of-type:pl-0">
                  {them.nom}
                </div>
              ))}
            </div>

            {/* Budget */}
            <NiveauBudget budget={budget} />
          </div>

          {/* Titre de l'action */}
          <h3>{titre}</h3>

          {/* Badge de complexité */}
          <Badge
            title={`Complexité ${valeurToBadge[complexite].nom}`}
            size="sm"
            state={valeurToBadge[complexite].style}
          />

          {/* Description */}
          <p className="paragraphe-18 mt-6">{description}</p>

          {/* Ressources externes */}
          {/* {!!ressources && (
            <div className="flex items-center gap-2">
              <Icon
                icon="external-link-line"
                size="lg"
                className="text-primary-4"
              />
              <a
                href={ressources}
                className="text-primary-10 font-bold bg-none active:!bg-transparent after:hidden border-b border-primary-10 hover:border-b-2 p-px hover:pb-0"
                target="_blank"
                rel="noreferrer noopener"
              >
                Consulter les ressources externes
              </a>
            </div>
          )} */}

          {/* Collectivités sur la même action */}
          {(!!nbCollectivitesEnCours || !!nbCollectivitesRealise) && (
            <div className="grid grid-cols-1 md:grid-cols-2 gap-x-8 gap-y-4 mt-6">
              {!!nbCollectivitesEnCours && (
                <div className="text-primary-9 text-sm font-medium border border-grey-3 rounded-md px-6 py-5 shadow">
                  <span className="font-bold">
                    {nbCollectivitesEnCours} collectivités
                  </span>{' '}
                  sont en cours sur cette action
                </div>
              )}

              {!!nbCollectivitesRealise && (
                <div className="text-primary-9 text-sm font-medium border border-grey-3 rounded-md px-6 py-5 shadow">
                  <span className="font-bold">
                    {nbCollectivitesRealise} collectivités
                  </span>{' '}
                  ont réalisé cette action
                </div>
              )}
            </div>
          )}
        </div>
      )}
      renderFooter={({close}) => (
        <ModalFooter variant="space">
          <ModalFooterSection>
            <Button
              variant="outlined"
              className={classNames({
                'bg-primary-2': statut === 'en_cours',
              })}
              onClick={() => {
                onUpdateStatus?.('en_cours');
                close();
              }}
            >
              En cours
            </Button>
            <Button
              variant="outlined"
              className={classNames({
                'bg-primary-2': statut === 'realise',
              })}
              onClick={() => {
                onUpdateStatus?.('realise');
                close();
              }}
            >
              Réalisée
            </Button>
          </ModalFooterSection>
          <Button
            icon={panier ? 'file-reduce-fill' : 'file-add-fill'}
            onClick={() => {
              onToggleSelected(!panier);
              close();
            }}
          >
            {panier ? 'Retirer du panier' : 'Ajouter au panier'}
          </Button>
        </ModalFooter>
      )}
    >
      {children}
    </Modal>
  );
};
