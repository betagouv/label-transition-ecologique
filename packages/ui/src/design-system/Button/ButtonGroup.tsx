import {CSSProperties} from 'react';
import classNames from 'classnames';

import {DefaultButtonProps, ButtonSize} from './types';
import {Button} from './Button';

type Props = {
  /** Listes de boutons à afficher avec leurs propriétés */
  buttons: DefaultButtonProps[];
  /** Permet d'ajuster les styles du container */
  className?: string;
  /** Taille des boutons */
  size?: ButtonSize;
  /** Id du bouton affiché dans une variante primary */
  activeButtonId?: string;
};

/**
 *  Affiche un groupe d'onglets.
 */
export const ButtonGroup = ({
  className,
  size = 'md',
  buttons,
  activeButtonId,
}: Props) => {
  /** Applique les bons styles pour les borders et border-radius
   * en fonction de la position du bouton dans le tableau */
  const getButtonStyles = (index: number): CSSProperties => {
    if (index === 0) {
      return {
        borderTopRightRadius: 0,
        borderBottomRightRadius: 0,
      };
    } else if (buttons.length === 2) {
      return {
        borderTopLeftRadius: 0,
        borderBottomLeftRadius: 0,
        borderLeft: 0,
      };
    } else {
      if (index > 0 && index < buttons.length - 1)
        return {
          borderRadius: 0,
          borderRight: 0,
          borderLeft: 0,
        };
      if (index === buttons.length - 1)
        return {
          borderTopLeftRadius: 0,
          borderBottomLeftRadius: 0,
        };
    }
  };

  return (
    <div className={classNames('flex items-center', className)}>
      {buttons.map((props, index) => (
        <Button
          key={index}
          {...props}
          size={size}
          variant={activeButtonId === props.id ? 'primary' : 'outlined'}
          className="!border-primary-7"
          style={getButtonStyles(index)}
        />
      ))}
    </div>
  );
};
