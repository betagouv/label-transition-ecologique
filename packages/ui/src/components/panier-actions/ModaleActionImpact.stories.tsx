import {Meta, StoryObj} from '@storybook/react';
import {ModaleActionImpact} from './ModaleActionImpact';
import {Button} from '@design-system/Button';

const meta: Meta<typeof ModaleActionImpact> = {
  title: "Components/Panier d'actions/Modale action à impact",
  component: ModaleActionImpact,
  argTypes: {
    complexite: {
      control: {type: 'select'},
    },
    budget: {
      control: {type: 'select'},
    },
  },
  args: {
    titre:
      'Désigner une (ou plusieurs personnes) chef de projet transition écologique',
    thematiques: [{id: 1, nom: 'Mieux consommer'}],
    complexite: 1,
    budget: 2,
    description:
      "Un comité de pilotage se réunit régulièrement (au moins 2 fois par an) pour s'assurer du bon avancement et de la priorisation de la programmation des actions de transition écologique. Ce comité de pilotage s'assure que les actions de transition écologique sont identifiées et priorisées dans l'exercice de programmation budgétaire et la programmation pluriannuelle d'investissement",
    ressources: 'https://www.territoiresentransitions.fr/',
    onToggleSelected: () => {},
    onUpdateStatus: () => {},
  },
};

export default meta;

type Story = StoryObj<typeof ModaleActionImpact>;

export const Default: Story = {
  render: args => (
    <ModaleActionImpact {...args}>
      <Button variant="outlined">Ouvrir la modale</Button>
    </ModaleActionImpact>
  ),
};
