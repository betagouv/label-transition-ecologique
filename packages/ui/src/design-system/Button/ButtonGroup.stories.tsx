import {useState} from 'react';
import {Meta, StoryObj} from '@storybook/react';

import {ButtonGroup} from './ButtonGroup';

const meta: Meta<typeof ButtonGroup> = {
  component: ButtonGroup,
  args: {
    activeButtonId: '2',
    buttons: [
      {id: '1', children: 'Bouton 1'},
      {id: '2', children: 'Bouton 2'},
      {id: '3', children: 'Bouton 3'},
    ],
  },
  render: args => {
    const [activeId, setActiveId] = useState(args.activeButtonId);
    return (
      <ButtonGroup
        {...args}
        activeButtonId={activeId}
        buttons={args.buttons.map(button => ({
          ...button,
          onClick: () => setActiveId(button.id),
        }))}
      />
    );
  },
};

export default meta;

type Story = StoryObj<typeof ButtonGroup>;

export const Default: Story = {
  args: {},
};

export const Size: Story = {
  args: {
    size: 'xs',
  },
};

export const Disabled: Story = {
  args: {
    buttons: [
      {id: '1', children: 'Bouton 1', disabled: true},
      {id: '2', children: 'Bouton 2'},
      {id: '3', children: 'Bouton 3'},
    ],
  },
};

export const Deux: Story = {
  args: {
    buttons: [
      {id: '1', children: 'Bouton 1'},
      {id: '2', children: 'Bouton 2'},
    ],
  },
};
