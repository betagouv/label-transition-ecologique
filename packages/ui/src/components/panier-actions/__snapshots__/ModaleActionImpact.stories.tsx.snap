// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components/Panier d'actions/Modale action Ã  impact Default smoke-test 1`] = `
<button aria-expanded="false"
        aria-haspopup="dialog"
        class="w-fit flex items-center border-solid group gap-2 cursor-pointer rounded-lg border font-bold text-primary-7 hover:!text-primary-8 bg-primary-0 hover:!bg-primary-1 border-primary-7 hover:!border-primary-8 text-base py-3 px-6"
>
  Ouvrir la modale
</button>
`;
